// pushfeedback@0.1.82/dist/collection/components/feedback-button/feedback-button.js
import { Host, h } from '@stencil/core';
export class FeedbackButton {
  constructor() {
    this.buttonPosition = 'default';
    this.buttonStyle = 'default';
    this.hideIcon = false;
    this.hideMobile = false;
    this.sessionId = '';
    this.metadata = '';
    this.submit = false;
    this.customFont = false;
    this.emailAddress = '';
    this.isEmailRequired = false;
    this.fetchData = true;
    this.hideEmail = false;
    this.hidePrivacyPolicy = true;
    this.hideRating = false;
    this.hideScreenshotButton = false;
    this.modalPosition = 'center';
    this.project = '';
    this.rating = undefined;
    this.ratingMode = 'thumbs';
    this.emailPlaceholder = 'Email address (optional)';
    this.errorMessage = 'Please try again later.';
    this.errorMessage403 = 'The request URL does not match the one defined in PushFeedback for this project.';
    this.errorMessage404 = 'We could not find the provided project id in PushFeedback.';
    this.footerText = '';
    this.messagePlaceholder = 'Share your thoughts...';
    this.modalTitle = 'Share your feedback';
    this.modalTitleError = 'Oops!';
    this.modalTitleSuccess = 'Thanks for your feedback!';
    this.privacyPolicyText = "I have read and expressly consent to the terms of the <a href='https://pushfeedback.com/privacy'>Privacy Policy</a>.";
    this.ratingPlaceholder = 'Was this page helpful?';
    this.ratingStarsPlaceholder = 'How would you rate this page?';
    this.sendButtonText = 'Send';
    this.successMessage = '';
    this.recaptchaText = 'This form is protected by reCAPTCHA and the Google <a href="https://policies.google.com/privacy" target="_blank" rel="noopener noreferrer">Privacy Policy</a> and <a href="https://policies.google.com/terms" target="_blank" rel="noopener noreferrer">Terms of Service</a> apply.';
    this.screenshotAttachedText = 'Screenshot attached';
    this.screenshotButtonText = 'Add a screenshot';
    this.screenshotTakingText = 'Taking screenshot...';
    this.screenshotEditTextButtonText = 'Edit text';
    this.screenshotEditorTitle = 'Edit screenshot';
    this.screenshotEditorCancelText = 'Cancel';
    this.screenshotEditorSaveText = 'Save';
    this.screenshotSizeLabelText = 'Size:';
    this.screenshotBorderLabelText = 'Border:';
    this.screenshotEditTextPromptText = 'Edit text:';
    this.screenshotErrorGeneral = 'Failed to capture screenshot.';
    this.screenshotErrorPermission = 'Permission denied. Please allow screen sharing to take screenshots.';
    this.screenshotErrorNotSupported = 'Screen capture is not supported in this browser.';
    this.screenshotErrorNotFound = 'No screen sources available for capture.';
    this.screenshotErrorCancelled = 'Screenshot capture was cancelled.';
    this.screenshotErrorBrowserNotSupported = 'Your browser does not support screen capture. Please use a browser like Chrome, Firefox, or Safari on desktop.';
    this.screenshotErrorUnexpected = 'An unexpected error occurred. Please try again.';
  }
  componentWillLoad() {
    if (!this.sessionId) {
      let storedSessionId = localStorage.getItem('pushfeedback_sessionid');
      if (!storedSessionId) {
        storedSessionId = this.generateRandomSessionId();
        localStorage.setItem('pushfeedback_sessionid', storedSessionId);
        this.sessionId = storedSessionId;
      }
    }
    else {
      localStorage.setItem('pushfeedback_sessionid', this.sessionId);
    }
  }
  componentDidLoad() {
    if (this.buttonPosition === 'center-right') {
      const buttonContent = this.el.shadowRoot.querySelector('.feedback-button-content');
      let adjustement = 0;
      if (this.isSafariBrowser()) {
        adjustement = 5;
      }
      buttonContent.style.right = `${((buttonContent.offsetWidth + adjustement) / 2) * -1}px`;
    }
    if (!this.customFont) {
      this.loadInterFont();
    }
  }
  connectedCallback() {
    this.feedbackModal = document.createElement('feedback-modal');
    const props = [
      'customFont',
      'emailAddress',
      'fetchData',
      'hideEmail',
      'hidePrivacyPolicy',
      'hideRating',
      'hideScreenshotButton',
      'isEmailRequired',
      'modalPosition',
      'project',
      'rating',
      'ratingMode',
      'screenshotEditorTitle',
      'screenshotEditorCancelText',
      'screenshotEditorSaveText',
      'screenshotEditTextButtonText',
      'screenshotSizeLabelText',
      'screenshotBorderLabelText',
      'screenshotEditTextPromptText',
      'screenshotErrorGeneral',
      'screenshotErrorPermission',
      'screenshotErrorNotSupported',
      'screenshotErrorNotFound',
      'screenshotErrorCancelled',
      'screenshotErrorBrowserNotSupported',
      'screenshotErrorUnexpected',
      'emailPlaceholder',
      'errorMessage',
      'errorMessage403',
      'errorMessage404',
      'footerText',
      'messagePlaceholder',
      'metadata',
      'modalTitle',
      'modalTitleError',
      'modalTitleSuccess',
      'privacyPolicyText',
      'ratingPlaceholder',
      'ratingStarsPlaceholder',
      'screenshotAttachedText',
      'screenshotButtonText',
      'screenshotTakingText',
      'sendButtonText',
      'successMessage',
      'recaptchaText',
    ];
    props.forEach((prop) => {
      this.feedbackModal[prop] = this[prop];
    });
    document.body.appendChild(this.feedbackModal);
  }
  disconnectedCallback() {
    document.body.removeChild(this.feedbackModal);
  }
  generateRandomSessionId(length = 16) {
    return Math.random().toString(36).substr(2, length);
  }
  isSafariBrowser() {
    const isSafari = /safari/i.test(navigator.userAgent) && !/chrome/i.test(navigator.userAgent);
    return isSafari;
  }
  loadInterFont() {
    const link = document.createElement('link');
    link.href = 'https://fonts.googleapis.com/css2?family=Inter:wght@400;500&display=swap';
    link.rel = 'stylesheet';
    document.head.appendChild(link);
  }
  showModal() {
    if (this.submit) {
      this.submitRatingFeedback();
    }
    else {
      this.feedbackModal.openModal();
    }
  }
  async submitRatingFeedback() {
    try {
      // Get reCAPTCHA token if enabled
      let recaptchaToken = null;
      if (this.feedbackModal.recaptchaEnabled) {
        recaptchaToken = await this.feedbackModal.getRecaptchaToken();
        if (!recaptchaToken) {
          // If reCAPTCHA is enabled but token retrieval failed, emit error
          const response = {
            status: 500,
            message: 'Failed to verify reCAPTCHA. Please try again.',
          };
          this.feedbackError.emit({ error: response });
          return;
        }
      }
      const body = Object.assign({ url: window.location.href, project: this.project, rating: this.rating || -1, ratingMode: this.ratingMode, message: '', metadata: this.metadata, session: localStorage.getItem('pushfeedback_sessionid') || '' }, (recaptchaToken && { recaptchaToken }));
      const res = await fetch('https://app.pushfeedback.com/api/feedback/', {
        method: 'POST',
        body: JSON.stringify(body),
        headers: {
          'Content-Type': 'application/json',
        },
      });
      if (res.status === 201) {
        const feedback_with_id = Object.assign(Object.assign({}, body), { id: await res.json() });
        this.feedbackSent.emit({ feedback: feedback_with_id });
      }
      else if (res.status === 202) {
        const limitExceededBody = {
          message: "You received a new feedback entry. You've reached the 25 message limit for your current plan. Upgrade to continue receiving feedback.",
          id: await res.json()
        };
        this.feedbackSent.emit({ feedback: limitExceededBody });
      }
      else {
        const errorText = await res.text();
        const response = {
          status: res.status,
          message: errorText,
        };
        this.feedbackError.emit({ error: response });
      }
    }
    catch (error) {
      const response = {
        status: 500,
        message: error,
      };
      this.feedbackError.emit({ error: response });
    }
  }
  render() {
    return (h(Host, null, h("a", { class: `feedback-button-content feedback-button-content--${this.buttonStyle} feedback-button-content--${this.buttonPosition} ${this.customFont ? 'feedback-button-content--custom-font' : ''} ${this.hideMobile ? 'feedback-button-content--hide-mobile' : ''}`, onClick: () => this.showModal() }, !this.hideIcon && this.buttonStyle != 'default' && (h("span", { class: "feedback-button-content-icon" }, h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "#fff", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", class: "icon-edit" }, h("path", { d: "M12 20h9" }), h("path", { d: "M16.5 3.5a2.121 2.121 0 0 1 3 3L7 19l-4 1 1-4L16.5 3.5z" })))), h("slot", null))));
  }
  static get is() { return "feedback-button"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["feedback-button.css"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["feedback-button.css"]
    };
  }
  static get properties() {
    return {
      "buttonPosition": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "button-position",
        "reflect": false,
        "defaultValue": "'default'"
      },
      "buttonStyle": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "button-style",
        "reflect": false,
        "defaultValue": "'default'"
      },
      "hideIcon": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "hide-icon",
        "reflect": false,
        "defaultValue": "false"
      },
      "hideMobile": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "hide-mobile",
        "reflect": false,
        "defaultValue": "false"
      },
      "sessionId": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "session-id",
        "reflect": true,
        "defaultValue": "''"
      },
      "metadata": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "metadata",
        "reflect": false,
        "defaultValue": "''"
      },
      "submit": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "submit",
        "reflect": false,
        "defaultValue": "false"
      },
      "customFont": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "custom-font",
        "reflect": false,
        "defaultValue": "false"
      },
      "emailAddress": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "email-address",
        "reflect": false,
        "defaultValue": "''"
      },
      "isEmailRequired": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "is-email-required",
        "reflect": false,
        "defaultValue": "false"
      },
      "fetchData": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "fetch-data",
        "reflect": false,
        "defaultValue": "true"
      },
      "hideEmail": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "hide-email",
        "reflect": false,
        "defaultValue": "false"
      },
      "hidePrivacyPolicy": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "hide-privacy-policy",
        "reflect": false,
        "defaultValue": "true"
      },
      "hideRating": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "hide-rating",
        "reflect": false,
        "defaultValue": "false"
      },
      "hideScreenshotButton": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "hide-screenshot-button",
        "reflect": false,
        "defaultValue": "false"
      },
      "modalPosition": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "modal-position",
        "reflect": false,
        "defaultValue": "'center'"
      },
      "project": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "project",
        "reflect": false,
        "defaultValue": "''"
      },
      "rating": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "rating",
        "reflect": false
      },
      "ratingMode": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "rating-mode",
        "reflect": false,
        "defaultValue": "'thumbs'"
      },
      "emailPlaceholder": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "email-placeholder",
        "reflect": false,
        "defaultValue": "'Email address (optional)'"
      },
      "errorMessage": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "error-message",
        "reflect": false,
        "defaultValue": "'Please try again later.'"
      },
      "errorMessage403": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "error-message-4-0-3",
        "reflect": false,
        "defaultValue": "'The request URL does not match the one defined in PushFeedback for this project.'"
      },
      "errorMessage404": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "error-message-4-0-4",
        "reflect": false,
        "defaultValue": "'We could not find the provided project id in PushFeedback.'"
      },
      "footerText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "footer-text",
        "reflect": false,
        "defaultValue": "''"
      },
      "messagePlaceholder": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "message-placeholder",
        "reflect": false,
        "defaultValue": "'Share your thoughts...'"
      },
      "modalTitle": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "modal-title",
        "reflect": false,
        "defaultValue": "'Share your feedback'"
      },
      "modalTitleError": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "modal-title-error",
        "reflect": false,
        "defaultValue": "'Oops!'"
      },
      "modalTitleSuccess": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "modal-title-success",
        "reflect": false,
        "defaultValue": "'Thanks for your feedback!'"
      },
      "privacyPolicyText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "privacy-policy-text",
        "reflect": false,
        "defaultValue": "\"I have read and expressly consent to the terms of the <a href='https://pushfeedback.com/privacy'>Privacy Policy</a>.\""
      },
      "ratingPlaceholder": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "rating-placeholder",
        "reflect": false,
        "defaultValue": "'Was this page helpful?'"
      },
      "ratingStarsPlaceholder": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "rating-stars-placeholder",
        "reflect": false,
        "defaultValue": "'How would you rate this page?'"
      },
      "sendButtonText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "send-button-text",
        "reflect": false,
        "defaultValue": "'Send'"
      },
      "successMessage": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "success-message",
        "reflect": false,
        "defaultValue": "''"
      },
      "recaptchaText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "recaptcha-text",
        "reflect": false,
        "defaultValue": "'This form is protected by reCAPTCHA and the Google <a href=\"https://policies.google.com/privacy\" target=\"_blank\" rel=\"noopener noreferrer\">Privacy Policy</a> and <a href=\"https://policies.google.com/terms\" target=\"_blank\" rel=\"noopener noreferrer\">Terms of Service</a> apply.'"
      },
      "screenshotAttachedText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-attached-text",
        "reflect": false,
        "defaultValue": "'Screenshot attached'"
      },
      "screenshotButtonText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-button-text",
        "reflect": false,
        "defaultValue": "'Add a screenshot'"
      },
      "screenshotTakingText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-taking-text",
        "reflect": false,
        "defaultValue": "'Taking screenshot...'"
      },
      "screenshotEditTextButtonText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-edit-text-button-text",
        "reflect": false,
        "defaultValue": "'Edit text'"
      },
      "screenshotEditorTitle": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-editor-title",
        "reflect": false,
        "defaultValue": "'Edit screenshot'"
      },
      "screenshotEditorCancelText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-editor-cancel-text",
        "reflect": false,
        "defaultValue": "'Cancel'"
      },
      "screenshotEditorSaveText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-editor-save-text",
        "reflect": false,
        "defaultValue": "'Save'"
      },
      "screenshotSizeLabelText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-size-label-text",
        "reflect": false,
        "defaultValue": "'Size:'"
      },
      "screenshotBorderLabelText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-border-label-text",
        "reflect": false,
        "defaultValue": "'Border:'"
      },
      "screenshotEditTextPromptText": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-edit-text-prompt-text",
        "reflect": false,
        "defaultValue": "'Edit text:'"
      },
      "screenshotErrorGeneral": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-error-general",
        "reflect": false,
        "defaultValue": "'Failed to capture screenshot.'"
      },
      "screenshotErrorPermission": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-error-permission",
        "reflect": false,
        "defaultValue": "'Permission denied. Please allow screen sharing to take screenshots.'"
      },
      "screenshotErrorNotSupported": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-error-not-supported",
        "reflect": false,
        "defaultValue": "'Screen capture is not supported in this browser.'"
      },
      "screenshotErrorNotFound": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-error-not-found",
        "reflect": false,
        "defaultValue": "'No screen sources available for capture.'"
      },
      "screenshotErrorCancelled": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-error-cancelled",
        "reflect": false,
        "defaultValue": "'Screenshot capture was cancelled.'"
      },
      "screenshotErrorBrowserNotSupported": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-error-browser-not-supported",
        "reflect": false,
        "defaultValue": "'Your browser does not support screen capture. Please use a browser like Chrome, Firefox, or Safari on desktop.'"
      },
      "screenshotErrorUnexpected": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "screenshot-error-unexpected",
        "reflect": false,
        "defaultValue": "'An unexpected error occurred. Please try again.'"
      }
    };
  }
  static get events() {
    return [{
        "method": "feedbackSent",
        "name": "feedbackSent",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": ""
        },
        "complexType": {
          "original": "{ feedback: any }",
          "resolved": "{ feedback: any; }",
          "references": {}
        }
      }, {
        "method": "feedbackError",
        "name": "feedbackError",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": ""
        },
        "complexType": {
          "original": "{ error: any }",
          "resolved": "{ error: any; }",
          "references": {}
        }
      }];
  }
  static get elementRef() { return "el"; }
}
