// pushfeedback/dist/components/feedback-button.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { d as defineCustomElement$3 } from './canvas-editor2.js';
import { d as defineCustomElement$2 } from './feedback-modal2.js';

const feedbackButtonCss = ".feedback-button-content{cursor:pointer;max-width:fit-content;letter-spacing:var(--feedback-button-letter-spacing);z-index:var(--feedback-button-z-index);font-family:var(--feedback-font-family)}.feedback-button-content--custom-font{font-family:inherit}.feedback-button-content--light{align-items:center;background-color:var(--feedback-button-light-bg-color);border-radius:var(--feedback-button-border-radius);box-shadow:rgba(60, 64, 67, 0.3) 0px 1px 2px 0px, rgba(60, 64, 67, 0.15) 0px 2px 6px 2px;box-sizing:border-box;color:var(--feedback-button-light-text-color);display:flex;font-size:var(--feedback-button-text-font-size);font-weight:var(--feedback-button-text-font-weight);padding:8px 15px}.feedback-button-content--dark{align-items:center;background-color:var(--feedback-button-dark-bg-color);border-radius:var(--feedback-button-border-radius);box-shadow:rgba(60, 64, 67, 0.3) 0px 1px 2px 0px, rgba(60, 64, 67, 0.15) 0px 2px 6px 2px;box-sizing:border-box;color:var(--feedback-button-dark-text-color);display:flex;font-weight:var(--feedback-button-text-font-weight);font-size:var(--feedback-button-text-font-size);padding:8px 15px}.icon-edit{stroke:var(--feedback-button-light-icon-color)}.feedback-button-content--dark .icon-edit{stroke:var(--feedback-button-dark-icon-color)}.feedback-button-content--bottom-right{bottom:10px;position:fixed;right:10px}.feedback-button-content--center-right{position:fixed;transform:rotate(-90deg) translateY(-50%);top:50%}.feedback-button-content--center-right.feedback-button-content--dark,.feedback-button-content--center-right.feedback-button-content--light{border-radius:4px;border-bottom-left-radius:0px;border-bottom-right-radius:0px}.feedback-button-content-icon{height:16px;margin-right:5px;width:16px}.feedback-button-content--center-right .feedback-button-content-icon{rotate:90deg}@media screen and (max-width: 767px){.feedback-button-content--hide-mobile{display:none}}";

const FeedbackButton$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.feedbackSent = createEvent(this, "feedbackSent", 7);
    this.feedbackError = createEvent(this, "feedbackError", 7);
    this.buttonPosition = 'default';
    this.buttonStyle = 'default';
    this.hideIcon = false;
    this.hideMobile = false;
    this.sessionId = '';
    this.metadata = '';
    this.submit = false;
    this.customFont = false;
    this.emailAddress = '';
    this.isEmailRequired = false;
    this.fetchData = true;
    this.hideEmail = false;
    this.hidePrivacyPolicy = true;
    this.hideRating = false;
    this.hideScreenshotButton = false;
    this.modalPosition = 'center';
    this.project = '';
    this.rating = undefined;
    this.ratingMode = 'thumbs';
    this.emailPlaceholder = 'Email address (optional)';
    this.errorMessage = 'Please try again later.';
    this.errorMessage403 = 'The request URL does not match the one defined in PushFeedback for this project.';
    this.errorMessage404 = 'We could not find the provided project id in PushFeedback.';
    this.footerText = '';
    this.messagePlaceholder = 'Share your thoughts...';
    this.modalTitle = 'Share your feedback';
    this.modalTitleError = 'Oops!';
    this.modalTitleSuccess = 'Thanks for your feedback!';
    this.privacyPolicyText = "I have read and expressly consent to the terms of the <a href='https://pushfeedback.com/privacy'>Privacy Policy</a>.";
    this.ratingPlaceholder = 'Was this page helpful?';
    this.ratingStarsPlaceholder = 'How would you rate this page?';
    this.sendButtonText = 'Send';
    this.successMessage = '';
    this.recaptchaText = 'This form is protected by reCAPTCHA and the Google <a href="https://policies.google.com/privacy" target="_blank" rel="noopener noreferrer">Privacy Policy</a> and <a href="https://policies.google.com/terms" target="_blank" rel="noopener noreferrer">Terms of Service</a> apply.';
    this.screenshotAttachedText = 'Screenshot attached';
    this.screenshotButtonText = 'Add a screenshot';
    this.screenshotTakingText = 'Taking screenshot...';
    this.screenshotEditTextButtonText = 'Edit text';
    this.screenshotEditorTitle = 'Edit screenshot';
    this.screenshotEditorCancelText = 'Cancel';
    this.screenshotEditorSaveText = 'Save';
    this.screenshotSizeLabelText = 'Size:';
    this.screenshotBorderLabelText = 'Border:';
    this.screenshotEditTextPromptText = 'Edit text:';
    this.screenshotErrorGeneral = 'Failed to capture screenshot.';
    this.screenshotErrorPermission = 'Permission denied. Please allow screen sharing to take screenshots.';
    this.screenshotErrorNotSupported = 'Screen capture is not supported in this browser.';
    this.screenshotErrorNotFound = 'No screen sources available for capture.';
    this.screenshotErrorCancelled = 'Screenshot capture was cancelled.';
    this.screenshotErrorBrowserNotSupported = 'Your browser does not support screen capture. Please use a browser like Chrome, Firefox, or Safari on desktop.';
    this.screenshotErrorUnexpected = 'An unexpected error occurred. Please try again.';
  }
  componentWillLoad() {
    if (!this.sessionId) {
      let storedSessionId = localStorage.getItem('pushfeedback_sessionid');
      if (!storedSessionId) {
        storedSessionId = this.generateRandomSessionId();
        localStorage.setItem('pushfeedback_sessionid', storedSessionId);
        this.sessionId = storedSessionId;
      }
    }
    else {
      localStorage.setItem('pushfeedback_sessionid', this.sessionId);
    }
  }
  componentDidLoad() {
    if (this.buttonPosition === 'center-right') {
      const buttonContent = this.el.shadowRoot.querySelector('.feedback-button-content');
      let adjustement = 0;
      if (this.isSafariBrowser()) {
        adjustement = 5;
      }
      buttonContent.style.right = `${((buttonContent.offsetWidth + adjustement) / 2) * -1}px`;
    }
    if (!this.customFont) {
      this.loadInterFont();
    }
  }
  connectedCallback() {
    this.feedbackModal = document.createElement('feedback-modal');
    const props = [
      'customFont',
      'emailAddress',
      'fetchData',
      'hideEmail',
      'hidePrivacyPolicy',
      'hideRating',
      'hideScreenshotButton',
      'isEmailRequired',
      'modalPosition',
      'project',
      'rating',
      'ratingMode',
      'screenshotEditorTitle',
      'screenshotEditorCancelText',
      'screenshotEditorSaveText',
      'screenshotEditTextButtonText',
      'screenshotSizeLabelText',
      'screenshotBorderLabelText',
      'screenshotEditTextPromptText',
      'screenshotErrorGeneral',
      'screenshotErrorPermission',
      'screenshotErrorNotSupported',
      'screenshotErrorNotFound',
      'screenshotErrorCancelled',
      'screenshotErrorBrowserNotSupported',
      'screenshotErrorUnexpected',
      'emailPlaceholder',
      'errorMessage',
      'errorMessage403',
      'errorMessage404',
      'footerText',
      'messagePlaceholder',
      'metadata',
      'modalTitle',
      'modalTitleError',
      'modalTitleSuccess',
      'privacyPolicyText',
      'ratingPlaceholder',
      'ratingStarsPlaceholder',
      'screenshotAttachedText',
      'screenshotButtonText',
      'screenshotTakingText',
      'sendButtonText',
      'successMessage',
      'recaptchaText',
    ];
    props.forEach((prop) => {
      this.feedbackModal[prop] = this[prop];
    });
    document.body.appendChild(this.feedbackModal);
  }
  disconnectedCallback() {
    document.body.removeChild(this.feedbackModal);
  }
  generateRandomSessionId(length = 16) {
    return Math.random().toString(36).substr(2, length);
  }
  isSafariBrowser() {
    const isSafari = /safari/i.test(navigator.userAgent) && !/chrome/i.test(navigator.userAgent);
    return isSafari;
  }
  loadInterFont() {
    const link = document.createElement('link');
    link.href = 'https://fonts.googleapis.com/css2?family=Inter:wght@400;500&display=swap';
    link.rel = 'stylesheet';
    document.head.appendChild(link);
  }
  showModal() {
    if (this.submit) {
      this.submitRatingFeedback();
    }
    else {
      this.feedbackModal.openModal();
    }
  }
  async submitRatingFeedback() {
    try {
      // Get reCAPTCHA token if enabled
      let recaptchaToken = null;
      if (this.feedbackModal.recaptchaEnabled) {
        recaptchaToken = await this.feedbackModal.getRecaptchaToken();
        if (!recaptchaToken) {
          // If reCAPTCHA is enabled but token retrieval failed, emit error
          const response = {
            status: 500,
            message: 'Failed to verify reCAPTCHA. Please try again.',
          };
          this.feedbackError.emit({ error: response });
          return;
        }
      }
      const body = Object.assign({ url: window.location.href, project: this.project, rating: this.rating || -1, ratingMode: this.ratingMode, message: '', metadata: this.metadata, session: localStorage.getItem('pushfeedback_sessionid') || '' }, (recaptchaToken && { recaptchaToken }));
      const res = await fetch('https://app.pushfeedback.com/api/feedback/', {
        method: 'POST',
        body: JSON.stringify(body),
        headers: {
          'Content-Type': 'application/json',
        },
      });
      if (res.status === 201) {
        const feedback_with_id = Object.assign(Object.assign({}, body), { id: await res.json() });
        this.feedbackSent.emit({ feedback: feedback_with_id });
      }
      else if (res.status === 202) {
        const limitExceededBody = {
          message: "You received a new feedback entry. You've reached the 25 message limit for your current plan. Upgrade to continue receiving feedback.",
          id: await res.json()
        };
        this.feedbackSent.emit({ feedback: limitExceededBody });
      }
      else {
        const errorText = await res.text();
        const response = {
          status: res.status,
          message: errorText,
        };
        this.feedbackError.emit({ error: response });
      }
    }
    catch (error) {
      const response = {
        status: 500,
        message: error,
      };
      this.feedbackError.emit({ error: response });
    }
  }
  render() {
    return (h(Host, null, h("a", { class: `feedback-button-content feedback-button-content--${this.buttonStyle} feedback-button-content--${this.buttonPosition} ${this.customFont ? 'feedback-button-content--custom-font' : ''} ${this.hideMobile ? 'feedback-button-content--hide-mobile' : ''}`, onClick: () => this.showModal() }, !this.hideIcon && this.buttonStyle != 'default' && (h("span", { class: "feedback-button-content-icon" }, h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "#fff", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", class: "icon-edit" }, h("path", { d: "M12 20h9" }), h("path", { d: "M16.5 3.5a2.121 2.121 0 0 1 3 3L7 19l-4 1 1-4L16.5 3.5z" })))), h("slot", null))));
  }
  get el() { return this; }
  static get style() { return feedbackButtonCss; }
}, [1, "feedback-button", {
    "buttonPosition": [1, "button-position"],
    "buttonStyle": [1, "button-style"],
    "hideIcon": [4, "hide-icon"],
    "hideMobile": [4, "hide-mobile"],
    "sessionId": [1537, "session-id"],
    "metadata": [1],
    "submit": [4],
    "customFont": [4, "custom-font"],
    "emailAddress": [1, "email-address"],
    "isEmailRequired": [4, "is-email-required"],
    "fetchData": [4, "fetch-data"],
    "hideEmail": [4, "hide-email"],
    "hidePrivacyPolicy": [4, "hide-privacy-policy"],
    "hideRating": [4, "hide-rating"],
    "hideScreenshotButton": [4, "hide-screenshot-button"],
    "modalPosition": [1, "modal-position"],
    "project": [1],
    "rating": [2],
    "ratingMode": [1, "rating-mode"],
    "emailPlaceholder": [1, "email-placeholder"],
    "errorMessage": [1, "error-message"],
    "errorMessage403": [1, "error-message-4-0-3"],
    "errorMessage404": [1, "error-message-4-0-4"],
    "footerText": [1, "footer-text"],
    "messagePlaceholder": [1, "message-placeholder"],
    "modalTitle": [1, "modal-title"],
    "modalTitleError": [1, "modal-title-error"],
    "modalTitleSuccess": [1, "modal-title-success"],
    "privacyPolicyText": [1, "privacy-policy-text"],
    "ratingPlaceholder": [1, "rating-placeholder"],
    "ratingStarsPlaceholder": [1, "rating-stars-placeholder"],
    "sendButtonText": [1, "send-button-text"],
    "successMessage": [1, "success-message"],
    "recaptchaText": [1, "recaptcha-text"],
    "screenshotAttachedText": [1, "screenshot-attached-text"],
    "screenshotButtonText": [1, "screenshot-button-text"],
    "screenshotTakingText": [1, "screenshot-taking-text"],
    "screenshotEditTextButtonText": [1, "screenshot-edit-text-button-text"],
    "screenshotEditorTitle": [1, "screenshot-editor-title"],
    "screenshotEditorCancelText": [1, "screenshot-editor-cancel-text"],
    "screenshotEditorSaveText": [1, "screenshot-editor-save-text"],
    "screenshotSizeLabelText": [1, "screenshot-size-label-text"],
    "screenshotBorderLabelText": [1, "screenshot-border-label-text"],
    "screenshotEditTextPromptText": [1, "screenshot-edit-text-prompt-text"],
    "screenshotErrorGeneral": [1, "screenshot-error-general"],
    "screenshotErrorPermission": [1, "screenshot-error-permission"],
    "screenshotErrorNotSupported": [1, "screenshot-error-not-supported"],
    "screenshotErrorNotFound": [1, "screenshot-error-not-found"],
    "screenshotErrorCancelled": [1, "screenshot-error-cancelled"],
    "screenshotErrorBrowserNotSupported": [1, "screenshot-error-browser-not-supported"],
    "screenshotErrorUnexpected": [1, "screenshot-error-unexpected"]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["feedback-button", "canvas-editor", "feedback-modal"];
  components.forEach(tagName => { switch (tagName) {
    case "feedback-button":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, FeedbackButton$1);
      }
      break;
    case "canvas-editor":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "feedback-modal":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const FeedbackButton = FeedbackButton$1;
const defineCustomElement = defineCustomElement$1;

export { FeedbackButton, defineCustomElement };
