// pushfeedback/dist/components/feedback-modal2.js
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { d as defineCustomElement$1 } from './canvas-editor2.js';

const feedbackModalCss = ".text-center{flex-grow:1;text-align:center}.feedback-modal-wrapper *{font-family:var(--feedback-font-family)}.feedback-modal-wrapper--custom-font *{font-family:inherit}.feedback-modal-wrapper{position:absolute;z-index:var(--feedback-modal-modal-wrapper-z-index)}.feedback-overlay{background-color:var(--feedback-modal-screenshot-bg-color);height:100%;left:0;opacity:0;position:fixed;top:0;width:100%;z-index:var(--feedback-modal-screnshot-z-index);transition:opacity 0.25s cubic-bezier(0.4, 0, 0.2, 1);backdrop-filter:blur(4px);-webkit-backdrop-filter:blur(4px)}.feedback-overlay--visible{opacity:1}.feedback-modal{display:inline-block;position:relative}.feedback-modal-content{background-color:var(--feedback-modal-content-bg-color);border:1px solid rgba(0, 0, 0, 0.08);border-radius:var(--feedback-modal-content-border-radius);box-shadow:0px 0px 0px 1px rgba(0, 0, 0, 0.02),\n    0px 2px 4px rgba(0, 0, 0, 0.04),\n    0px 8px 16px rgba(0, 0, 0, 0.06),\n    0px 16px 32px rgba(0, 0, 0, 0.04);box-sizing:border-box;color:var(--feedback-modal-content-text-color);display:flex;flex-direction:column;left:50%;max-width:90%;padding:24px;position:fixed;top:50%;transform:translate(-50%, -50%) scale(0.96);opacity:0;width:100%;z-index:var(--feedback-modal-content-z-index);transition:transform 0.25s cubic-bezier(0.4, 0, 0.2, 1), opacity 0.25s cubic-bezier(0.4, 0, 0.2, 1)}.feedback-modal-content--open{transform:translate(-50%, -50%) scale(1);opacity:1}.feedback-modal-header{color:var(--feedback-modal-header-text-color);display:flex;font-size:var(--feedback-modal-header-font-size);font-weight:var(--feedback-modal-header-font-weight);justify-content:space-between;align-items:flex-start;margin-bottom:24px;letter-spacing:-0.01em}.feedback-modal-header-content{flex:1}.feedback-modal-header-text{display:flex;flex-direction:column;gap:6px}.feedback-modal-title{display:block}.feedback-modal-powered-by{display:block;font-size:13px;font-weight:400;color:var(--feedback-color-gray-500);line-height:1.4}.feedback-modal-powered-by a{color:var(--feedback-modal-header-text-color);text-decoration:none;font-weight:500}.feedback-modal-powered-by a:hover{text-decoration:underline}.feedback-modal-header--no-content{margin-bottom:0}.feedback-modal-rating-buttons{width:100%;margin-bottom:24px;display:flex;gap:4px}.feedback-modal-rating-button{padding:0;background-color:transparent;border:transparent;margin-right:8px;cursor:pointer}.feedback-modal-rating-buttons--stars .feedback-modal-rating-button{transition:transform 0.15s ease, opacity 0.15s ease}.feedback-modal-rating-buttons--stars .feedback-modal-rating-button:hover{transform:scale(1.1)}.feedback-modal-rating-buttons--thumbs .feedback-modal-rating-button{border:1.5px solid var(--feedback-modal-button-border-color);border-radius:var(--feedback-modal-button-border-radius);color:var(--feedback-modal-button-text-color);font-size:var(--feedback-modal-button-font-size);font-weight:500;margin-right:12px;justify-content:center;padding:10px 16px;transition:all 0.2s cubic-bezier(0.4, 0, 0.2, 1);box-shadow:0 1px 2px rgba(0, 0, 0, 0.04)}.feedback-modal-rating-buttons--thumbs .feedback-modal-rating-button:hover{transform:translateY(-2px);box-shadow:0 4px 8px rgba(0, 0, 0, 0.12)}.feedback-modal-rating-buttons--thumbs .feedback-modal-rating-button:hover,.feedback-modal-rating-buttons--thumbs .feedback-modal-rating-button--selected{background-color:var(--feedback-modal-button-bg-color-active);border:1.5px solid var(--feedback-modal-button-border-color-active);color:var(--feedback-modal-button-text-color-active)}.feedback-modal-rating-buttons--thumbs .feedback-modal-rating-button:hover svg,.feedback-modal-rating-buttons--thumbs .feedback-modal-rating-button--selected svg{stroke:var(--feedback-modal-rating-button-selected-color)}.feedback-modal-rating-buttons svg{stroke:var(--feedback-modal-rating-button-color);cursor:pointer}.feedback-modal-rating-buttons--stars .feedback-modal-rating-button--selected svg{fill:var(--feedback-modal-rating-button-stars-selected-color);stroke:var(--feedback-modal-rating-button-stars-selected-color)}.feedback-modal-text textarea{background-color:var(--feedback-modal-input-bg-color);border:1.5px solid var(--feedback-modal-input-border-color);border-radius:var(--feedback-modal-input-border-radius);box-sizing:border-box;color:var(--feedback-modal-input-text-color);font-size:var(--feedback-modal-input-font-size);margin-bottom:20px;height:100px;min-height:100px;padding:12px;resize:vertical;width:100%;transition:border-color 0.2s ease, box-shadow 0.2s ease;line-height:1.5}.feedback-modal-text textarea:hover{border-color:var(--feedback-modal-input-border-color-hover)}.feedback-modal-email input{background-color:var(--feedback-modal-input-bg-color);border:1.5px solid var(--feedback-modal-input-border-color);border-radius:var(--feedback-modal-input-border-radius);box-sizing:border-box;color:var(--feedback-modal-input-text-color);font-size:var(--feedback-modal-input-font-size);margin-bottom:20px;height:44px;padding:12px;width:100%;transition:border-color 0.2s ease, box-shadow 0.2s ease}.feedback-modal-email input:hover{border-color:var(--feedback-modal-input-border-color-hover)}.feedback-modal-text textarea:-webkit-autofill,.feedback-modal-email input:-webkit-autofill,.feedback-modal-email input:-webkit-autofill:hover,.feedback-modal-email input:-webkit-autofill:focus,.feedback-modal-email input:-webkit-autofill:active{-webkit-box-shadow:0 0 0 1000px var(--feedback-modal-input-bg-color) inset !important;-webkit-text-fill-color:var(--feedback-modal-input-text-color) !important;transition:background-color 5000s ease-in-out 0s}.feedback-modal-privacy{font-size:var(--feedback-modal-input-font-size);margin-bottom:20px}.feedback-modal-text textarea:focus,.feedback-modal-email input:focus{border:1.5px solid var(--feedback-modal-input-border-color-focused);outline:none;box-shadow:0 0 0 3px rgba(59, 130, 246, 0.1)}.feedback-modal-buttons{display:flex;flex-direction:column}.feedback-modal-buttons .feedback-modal-button{margin-bottom:20px}.feedback-modal-button{align-items:center;background-color:transparent;border:1.5px solid var(--feedback-modal-button-border-color);border-radius:var(--feedback-modal-button-border-radius);color:var(--feedback-modal-button-text-color);cursor:pointer;display:flex;font-size:var(--feedback-modal-button-font-size);font-weight:500;justify-content:center;min-height:44px;padding:10px 16px;transition:all 0.2s cubic-bezier(0.4, 0, 0.2, 1);box-shadow:0 1px 2px rgba(0, 0, 0, 0.04)}.feedback-modal-button svg{margin-right:6px}.feedback-modal-button path{fill:var(--feedback-modal-button-icon-color)}.feedback-modal-button:hover path,.feedback-modal-button--active path{fill:var(--feedback-modal-button-icon-color-active)}.feedback-modal-button--submit{background-color:var(--feedback-modal-button-submit-bg-color);border:1.5px solid var(--feedback-modal-button-border-color-active);color:var(--feedback-modal-button-submit-text-color);box-shadow:0 2px 4px rgba(0, 0, 0, 0.1)}.feedback-modal-button:hover,.feedback-modal-button--active{background-color:var(--feedback-modal-button-bg-color-active);border:1.5px solid var(--feedback-modal-button-border-color-active);color:var(--feedback-modal-button-text-color-active);transform:translateY(-1px);box-shadow:0 4px 8px rgba(0, 0, 0, 0.12)}.feedback-modal-button--submit:hover{background-color:var(--feedback-modal-button-submit-bg-color-hover);border:1.5px solid var(--feedback-modal-button-submit-border-color-hover);color:var(--feedback-modal-button-submit-text-color-hover);transform:translateY(-1px);box-shadow:0 6px 12px rgba(0, 0, 0, 0.15)}.feedback-modal-button--submit:active{transform:translateY(0);box-shadow:0 2px 4px rgba(0, 0, 0, 0.1)}.feedback-modal-input-heading{display:block;font-size:14px;font-weight:500;padding-bottom:12px;color:var(--feedback-modal-header-text-color);letter-spacing:-0.01em}.feedback-modal-footer{font-size:12px;text-align:center}.feedback-modal-footer a{color:var(--feedback-modal-footer-link);font-weight:500;text-decoration:none}.feedback-logo,.feedback-footer-text{display:block;text-align:center;margin-top:5px}.feedback-footer-text{margin-top:10px;line-height:1.5}.feedback-footer-combined{display:block;text-align:center;font-size:11px;color:#666;line-height:1.5}.feedback-footer-combined a{color:#666;text-decoration:underline}.feedback-recaptcha-notice{display:block;text-align:center;margin-bottom:10px;font-size:11px;color:#666;line-height:1.4}.feedback-recaptcha-notice a{color:#666;text-decoration:underline}.feedback-modal-close{background-color:var(--feedback-modal-close-bg-color);border:0;border-radius:6px;cursor:pointer;height:32px;width:32px;margin-left:auto;padding:0;display:flex;align-items:center;justify-content:center;transition:all 0.2s ease}.feedback-modal-close:hover{background-color:rgba(0, 0, 0, 0.06);transform:scale(1.05)}.feedback-modal-close:active{transform:scale(0.95)}.feedback-modal-close svg{stroke:var(--feedback-modal-close-color)}.feedback-modal-screenshot{background-color:var(--feedback-modal-screenshot-bg-color);height:100%;left:0;position:fixed;top:0;width:100%;z-index:var(--feedback-modal-screnshot-z-index)}.feedback-modal-screenshot-header{align-items:center;background-color:var(--feedback-modal-screenshot-header-bg-color);border-radius:var(--feedback-modal-content-border-radius);box-shadow:0px 1px 2px 0px rgba(60, 64, 67, .30), 0px 2px 6px 2px rgba(60, 64, 67, .15);box-sizing:border-box;color:var(--feedback-modal-screenshot-header-text-color);cursor:pointer;display:flex;left:50%;top:20px;transform:translateX(-50%);padding:10px;position:fixed;width:max-content;z-index:var(--feedback-modal-screenshot-header-z-index)}.feedback-modal-screenshot-close{height:24px;padding-left:10px;width:24px}.feedback-modal-screenshot-close svg{stroke:var(--feedback-modal-close-color)}.feedback-modal-message{font-size:var(--feedback-modal-message-font-size);margin:0;line-height:1.6;color:var(--feedback-modal-content-text-color)}.feedback-modal-element-hover{background-color:transparent;cursor:pointer;border:1px solid var(--feedback-modal-element-hover-border-color)}.feedback-modal-element-selected{background-color:transparent;border:3px solid var(--feedback-modal-element-selected-border-color) !important;box-shadow:0 0 0 2px rgba(0, 123, 255, 0.3) !important}.screenshot-preview{display:inline-block;width:32px;height:32px;overflow:hidden;border-radius:6px;margin-right:10px;box-shadow:0 2px 4px rgba(0, 0, 0, 0.1);cursor:pointer;transition:all 0.2s cubic-bezier(0.4, 0, 0.2, 1);border:2px solid rgba(255, 255, 255, 0.8)}.screenshot-preview:hover{transform:scale(1.15);box-shadow:0 4px 8px rgba(0, 0, 0, 0.15)}.screenshot-preview img{width:100%;height:100%;object-fit:cover}.screenshot-loading{display:inline-flex;align-items:center;margin-right:8px}@media screen and (min-width: 768px){.feedback-modal-content{max-width:var(--feedback-modal-content-max-width)}.feedback-modal-content.feedback-modal-content--bottom-right{bottom:var(--feedback-modal-content-position-bottom);left:initial;right:var(--feedback-modal-content-position-right);top:initial;transform:initial}.feedback-modal-content.feedback-modal-content--bottom-left{bottom:var(--feedback-modal-content-position-bottom);left:var(--feedback-modal-content-position-left);top:initial;transform:initial}.feedback-modal-content.feedback-modal-content--top-right{right:var(--feedback-modal-content-position-right);top:var(--feedback-modal-content-position-top);transform:initial}.feedback-modal-content.feedback-modal-content--top-left{left:var(--feedback-modal-content-position-left);top:var(--feedback-modal-content-position-top);transform:initial}.feedback-modal-content.feedback-modal-content--center-left{left:5px;right:auto;top:50%;transform:translateY(-50%)}.feedback-modal-content.feedback-modal-content--center-right{left:auto;right:5px;top:50%;transform:translateY(-50%)}.feedback-modal-content.feedback-modal-content--sidebar-left.feedback-modal-content--open,.feedback-modal-content.feedback-modal-content--sidebar-right.feedback-modal-content--open{transform:translateX(0)}.feedback-modal-content.feedback-modal-content--sidebar-left{max-width:var(--feedback-modal-content-sidebar-max-width);left:0;right:auto;height:100vh;top:0;transform:translateX(-100%);transition:transform 0.5s ease-in-out;border-radius:0}.feedback-modal-content.feedback-modal-content--sidebar-right{max-width:var(--feedback-modal-content-sidebar-max-width);left:auto;right:0;height:100vh;top:0;transform:translateX(100%);transition:transform 0.5s ease-in-out;border-radius:0}.feedback-modal-text textarea{height:150px;min-height:150px}.feedback-modal-content.feedback-modal-content--bottom-right{transform:translateY(20px)}.feedback-modal-content.feedback-modal-content--bottom-right.feedback-modal-content--open{transform:translateY(0)}.feedback-modal-content.feedback-modal-content--bottom-left{transform:translateY(20px)}.feedback-modal-content.feedback-modal-content--bottom-left.feedback-modal-content--open{transform:translateY(0)}.feedback-modal-content.feedback-modal-content--top-right{transform:translateY(-20px)}.feedback-modal-content.feedback-modal-content--top-right.feedback-modal-content--open{transform:translateY(0)}.feedback-modal-content.feedback-modal-content--top-left{transform:translateY(-20px)}.feedback-modal-content.feedback-modal-content--top-left.feedback-modal-content--open{transform:translateY(0)}}@keyframes feather-spin{0%{transform:rotate(0deg)}100%{transform:rotate(360deg)}}.feather-loader{animation:feather-spin 1s linear infinite;display:block}.screenshot-error-notification{position:fixed;top:20px;left:50%;transform:translateX(-50%);z-index:10001;max-width:500px;width:90%;animation:slideDown 0.3s ease-out}@keyframes slideDown{from{opacity:0;transform:translateX(-50%) translateY(-20px)}to{opacity:1;transform:translateX(-50%) translateY(0)}}.screenshot-error-content{background:#fee;border:1.5px solid #fcc;border-radius:10px;padding:14px 18px;display:flex;align-items:center;gap:12px;box-shadow:0 0 0 1px rgba(197, 48, 48, 0.05),\n    0 4px 6px rgba(0, 0, 0, 0.07),\n    0 10px 20px rgba(0, 0, 0, 0.1);color:#c53030}.screenshot-error-content svg:first-child{color:#e53e3e;flex-shrink:0}.screenshot-error-content span{flex:1;font-size:14px;line-height:1.4;font-weight:500}.error-close-btn{background:none;border:none;cursor:pointer;padding:6px;border-radius:6px;color:#c53030;flex-shrink:0;transition:all 0.2s ease;display:flex;align-items:center;justify-content:center}.error-close-btn:hover{background:rgba(197, 48, 48, 0.15);transform:scale(1.1)}.error-close-btn:active{transform:scale(0.95)}@media screen and (max-width: 768px){.feedback-modal-content{width:100vw;height:100dvh;max-width:none;border-radius:0;top:0;left:0;transform:scale(0.95);padding:24px 20px;display:flex;flex-direction:column}.feedback-modal-content--open{transform:scale(1)}.feedback-modal-buttons{display:flex;flex-direction:column;gap:12px}.feedback-modal-button--screenshot{display:none !important}.feedback-modal-button--submit{width:100%}.feedback-modal-text textarea{flex:1;min-height:120px;resize:none}}.feedback-modal-wrapper--embedded{position:relative;z-index:auto;display:block;width:100%}.feedback-modal-content--embedded{position:relative !important;left:auto !important;right:auto !important;top:auto !important;bottom:auto !important;transform:none !important;opacity:1 !important;max-width:100%;margin:0 auto;box-shadow:0 2px 8px rgba(0, 0, 0, 0.1);text-align:left}.feedback-modal-content--embedded.feedback-modal-content--open{transform:none !important}@media screen and (max-width: 768px){.feedback-modal-content--embedded{width:100%;height:auto;border-radius:var(--feedback-modal-content-border-radius);padding:24px}.feedback-modal-content--embedded .feedback-modal-button--screenshot{display:flex !important}}";

const FeedbackModal = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.feedbackSent = createEvent(this, "feedbackSent", 7);
    this.feedbackError = createEvent(this, "feedbackError", 7);
    this.onScrollDebounced = () => {
      clearTimeout(this.scrollTimeout);
      this.scrollTimeout = setTimeout(() => {
        document.documentElement.classList.remove('feedback-modal-screenshot-closing');
        document.documentElement.style.top = '';
        window.removeEventListener('scroll', this.onScrollDebounced);
      }, 200);
    };
    this.handleSubmit = async (event) => {
      event.preventDefault();
      if (this.isEmailRequired && !this.formEmail) {
        return;
      }
      this.resetOverflow();
      this.showScreenshotMode = false;
      this.showScreenshotTopBar = false;
      this.showModal = false;
      this.sending = true;
      try {
        // Get reCAPTCHA token if enabled
        let recaptchaToken = null;
        if (this.recaptchaEnabled) {
          recaptchaToken = await this.getRecaptchaToken();
          if (!recaptchaToken) {
            // If reCAPTCHA is enabled but token retrieval failed, show error
            this.formSuccess = false;
            this.formError = true;
            this.formErrorStatus = 500;
            this.sending = false;
            this.showModal = true;
            return;
          }
        }
        const body = Object.assign({ url: window.location.href, message: this.formMessage, email: this.formEmail, project: this.project, screenshot: this.encodedScreenshot, rating: this.selectedRating, ratingMode: this.ratingMode, metadata: this.metadata, verification: this.formVerification, session: localStorage.getItem('pushfeedback_sessionid') || '' }, (recaptchaToken && { recaptchaToken }));
        const res = await fetch('https://app.pushfeedback.com/api/feedback/', {
          method: 'POST',
          body: JSON.stringify(body),
          headers: {
            'Content-Type': 'application/json',
          },
        });
        if (res.status === 201) {
          const feedback_with_id = Object.assign(Object.assign({}, body), { id: await res.json() });
          this.feedbackSent.emit({ feedback: feedback_with_id });
          this.formSuccess = true;
          this.formError = false;
        }
        else if (res.status === 202) {
          const limitExceededBody = {
            message: "You received a new feedback entry. You've reached the 25 message limit for your current plan. Upgrade to continue receiving feedback.",
            id: await res.json()
          };
          this.feedbackSent.emit({ feedback: limitExceededBody });
          this.formSuccess = true;
          this.formError = false;
        }
        else {
          const errorText = await res.text();
          const response = {
            status: res.status,
            message: errorText,
          };
          this.feedbackError.emit({ error: response });
          this.formSuccess = false;
          this.formError = true;
          this.formErrorStatus = res.status;
        }
      }
      catch (error) {
        const response = {
          status: 500,
          message: error,
        };
        this.feedbackError.emit({ error: response });
        this.formSuccess = false;
        this.formError = true;
        this.formErrorStatus = 500;
      }
      finally {
        this.sending = false;
        this.showModal = true;
      }
    };
    this.close = () => {
      this.isAnimating = false;
      setTimeout(() => {
        this.sending = false;
        this.showModal = false;
        this.showScreenshotMode = false;
        this.showScreenshotTopBar = false;
        this.hasSelectedElement = false;
        this.encodedScreenshot = null;
        // Remove highlight from ALL selected elements
        document.querySelectorAll('.feedback-modal-element-selected').forEach(el => {
          el.classList.remove('feedback-modal-element-selected');
        });
        // Reset form states
        this.formSuccess = false;
        this.formError = false;
        this.formErrorStatus = 500;
        this.formMessage = '';
        this.formEmail = '';
        this.resetOverflow();
      }, 200);
    };
    // Handle screenshot events from canvas editor
    this.handleScreenshotReady = (event) => {
      this.encodedScreenshot = event.detail.screenshot;
      this.showModal = true;
      this.takingScreenshot = false;
      this.showCanvasEditor = false;
      this.autoStartCapture = false;
    };
    this.handleScreenshotCancelled = () => {
      this.showModal = true;
      this.takingScreenshot = false;
      this.showCanvasEditor = false;
      this.autoStartCapture = false;
    };
    this.handleScreenshotError = (event) => {
      console.error('Screenshot error:', event.detail.error);
      // Store error message to display in feedback modal
      this.screenshotError = event.detail.error;
      this.showScreenshotError = true;
      // Close canvas editor and return to feedback modal
      this.showModal = true;
      this.takingScreenshot = false;
      this.showCanvasEditor = false;
      this.autoStartCapture = false;
      // Auto-hide error after 8 seconds
      setTimeout(() => {
        this.showScreenshotError = false;
      }, 8000);
    };
    // Trigger screenshot capture
    this.openScreenShot = () => {
      this.showModal = false;
      this.takingScreenshot = true;
      this.autoStartCapture = true; // Auto-start new screenshot
      this.showCanvasEditor = true;
    };
    // Open canvas editor for existing screenshot
    this.openCanvasEditor = (event) => {
      if (event) {
        event.stopPropagation();
      }
      this.showModal = false;
      this.autoStartCapture = false; // Don't auto-start, just edit existing
      this.showCanvasEditor = true;
    };
    this.sending = false;
    this.formMessage = '';
    this.formEmail = '';
    this.formSuccess = false;
    this.formVerification = '';
    this.formError = false;
    this.formErrorStatus = 500;
    this.encodedScreenshot = undefined;
    this.isPrivacyChecked = false;
    this.whitelabel = false;
    this.selectedRating = -1;
    this.overlayVisible = false;
    this.isAnimating = false;
    this.recaptchaEnabled = false;
    this.recaptchaSiteKey = null;
    this.takingScreenshot = false;
    this.showScreenshotError = false;
    this.screenshotError = '';
    this.showCanvasEditor = false;
    this.autoStartCapture = false;
    this.customFont = false;
    this.emailAddress = '';
    this.hideEmail = false;
    this.isEmailRequired = false;
    this.ratingMode = 'thumbs';
    this.hasSelectedElement = false;
    this.hidePrivacyPolicy = true;
    this.hideRating = false;
    this.hideScreenshotButton = false;
    this.project = '';
    this.showScreenshotMode = false;
    this.showScreenshotTopBar = false;
    this.showModal = false;
    this.rating = undefined;
    this.metadata = undefined;
    this.fetchData = true;
    this.embedded = false;
    this.emailPlaceholder = 'Email address (optional)';
    this.errorMessage = 'Please try again later.';
    this.errorMessage403 = 'The request URL does not match the one defined in PushFeedback for this project.';
    this.errorMessage404 = 'We could not find the provided project ID in PushFeedback.';
    this.messagePlaceholder = 'Share your thoughts...';
    this.footerText = '';
    this.modalPosition = 'center';
    this.modalTitle = 'Share your feedback';
    this.modalTitleError = 'Oops!';
    this.modalTitleSuccess = 'Thanks for your feedback!';
    this.privacyPolicyText = "I have read and expressly consent to the terms of the <a href='https://pushfeedback.com/privacy'>Privacy Policy</a>.";
    this.ratingPlaceholder = 'Was this page helpful?';
    this.ratingStarsPlaceholder = 'How would you rate this page?';
    this.sendButtonText = 'Send';
    this.successMessage = '';
    this.recaptchaText = 'This form is protected by reCAPTCHA and the Google <a href="https://policies.google.com/privacy" target="_blank" rel="noopener noreferrer">Privacy Policy</a> and <a href="https://policies.google.com/terms" target="_blank" rel="noopener noreferrer">Terms of Service</a> apply.';
    this.screenshotAttachedText = 'Screenshot attached';
    this.screenshotButtonText = 'Add a screenshot';
    this.screenshotTakingText = 'Taking screenshot...';
    this.screenshotEditTextButtonText = 'Edit text';
    this.screenshotEditorTitle = 'Edit screenshot';
    this.screenshotEditorCancelText = 'Cancel';
    this.screenshotEditorSaveText = 'Save';
    this.screenshotSizeLabelText = 'Size:';
    this.screenshotBorderLabelText = 'Border:';
    this.screenshotEditTextPromptText = 'Edit text:';
    this.screenshotErrorGeneral = 'Failed to capture screenshot.';
    this.screenshotErrorPermission = 'Permission denied. Please allow screen sharing to take screenshots.';
    this.screenshotErrorNotSupported = 'Screen capture is not supported in this browser.';
    this.screenshotErrorNotFound = 'No screen sources available for capture.';
    this.screenshotErrorCancelled = 'Screenshot capture was cancelled.';
    this.screenshotErrorBrowserNotSupported = 'Your browser does not support screen capture. Please use a browser like Chrome, Firefox, or Safari on desktop.';
    this.screenshotErrorUnexpected = 'An unexpected error occurred. Please try again.';
  }
  componentWillLoad() {
    if (this.fetchData)
      this.fetchProjectData();
    this.formEmail = this.emailAddress;
    if (this.rating) {
      this.selectedRating = this.rating;
    }
    if (this.ratingMode == 'thumbs' && this.rating == 0) {
      this.selectedRating = 5;
    }
    // Automatically show modal when embedded
    if (this.embedded) {
      this.showModal = true;
    }
  }
  async fetchProjectData() {
    try {
      const response = await fetch('https://app.pushfeedback.com/api/projects/' + this.project + '/');
      const data = await response.json();
      this.whitelabel = data.whitelabel;
      this.recaptchaEnabled = data.recaptcha_enabled || false;
      this.recaptchaSiteKey = data.recaptcha_site_key || null;
      // Load reCAPTCHA script if enabled
      if (this.recaptchaEnabled && this.recaptchaSiteKey) {
        this.loadRecaptchaScript();
      }
    }
    catch (error) {
      console.log(error);
    }
  }
  loadRecaptchaScript() {
    // Check if script already loaded
    if (document.querySelector('script[src*="google.com/recaptcha"]')) {
      return;
    }
    const script = document.createElement('script');
    script.src = `https://www.google.com/recaptcha/api.js?render=${this.recaptchaSiteKey}`;
    script.async = true;
    script.defer = true;
    document.head.appendChild(script);
  }
  async getRecaptchaToken() {
    if (!this.recaptchaEnabled || !this.recaptchaSiteKey) {
      return null;
    }
    try {
      // Wait for grecaptcha to be available
      let attempts = 0;
      while (!window['grecaptcha'] && attempts < 50) {
        await new Promise(resolve => setTimeout(resolve, 100));
        attempts++;
      }
      if (!window['grecaptcha']) {
        console.error('reCAPTCHA script not loaded');
        return null;
      }
      // Use ready() to ensure grecaptcha is fully initialized before executing
      return await new Promise((resolve, reject) => {
        window['grecaptcha'].ready(async () => {
          try {
            const token = await window['grecaptcha'].execute(this.recaptchaSiteKey, {
              action: 'submit_feedback'
            });
            resolve(token);
          }
          catch (error) {
            console.error('reCAPTCHA execution error:', error);
            reject(error);
          }
        });
      });
    }
    catch (error) {
      console.error('reCAPTCHA error:', error);
      return null;
    }
  }
  resetOverflow() {
    // Just clean up any stray classes, don't add/remove during screenshot
    document.documentElement.classList.remove('feedback-modal-screenshot-open');
    document.documentElement.classList.remove('feedback-modal-screenshot-open--scroll');
    document.documentElement.classList.remove('feedback-modal-screenshot-closing');
  }
  handleMessageInput(event) {
    this.formMessage = event.target.value;
  }
  handleEmailInput(event) {
    this.formEmail = event.target.value;
  }
  handleCheckboxChange(event) {
    this.isPrivacyChecked = event.target.checked;
  }
  handleVerification(event) {
    this.formVerification = event.target.value;
  }
  handleRatingChange(newRating) {
    this.selectedRating = newRating;
  }
  render() {
    return (h("div", { class: `feedback-modal-wrapper ${this.customFont ? 'feedback-modal-wrapper--custom-font' : ''} ${this.embedded ? 'feedback-modal-wrapper--embedded' : ''}` }, this.showCanvasEditor && (h("canvas-editor", { ref: (el) => this.canvasEditorRef = el, "canvas-editor-title": this.screenshotEditorTitle, "canvas-editor-cancel-text": this.screenshotEditorCancelText, "canvas-editor-save-text": this.screenshotEditorSaveText, "screenshot-taking-text": this.screenshotTakingText, "screenshot-attached-text": this.screenshotAttachedText, "screenshot-button-text": this.screenshotButtonText, "auto-start-screenshot": this.autoStartCapture, "existing-screenshot": this.encodedScreenshot || '', "edit-text-button-text": this.screenshotEditTextButtonText, "size-label-text": this.screenshotSizeLabelText, "border-label-text": this.screenshotBorderLabelText, "edit-text-prompt-text": this.screenshotEditTextPromptText, "screenshot-error-general": this.screenshotErrorGeneral, "screenshot-error-permission": this.screenshotErrorPermission, "screenshot-error-not-supported": this.screenshotErrorNotSupported, "screenshot-error-not-found": this.screenshotErrorNotFound, "screenshot-error-cancelled": this.screenshotErrorCancelled, "screenshot-error-browser-not-supported": this.screenshotErrorBrowserNotSupported, "screenshot-error-unexpected": this.screenshotErrorUnexpected, onScreenshotReady: this.handleScreenshotReady, onScreenshotCancelled: this.handleScreenshotCancelled, onScreenshotFailed: this.handleScreenshotError })), this.showScreenshotError && (h("div", { class: "screenshot-error-notification" }, h("div", { class: "screenshot-error-content" }, h("svg", { width: "20", height: "20", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round" }, h("circle", { cx: "12", cy: "12", r: "10" }), h("line", { x1: "15", y1: "9", x2: "9", y2: "15" }), h("line", { x1: "9", y1: "9", x2: "15", y2: "15" })), h("span", null, this.screenshotError), h("button", { class: "error-close-btn", onClick: () => this.showScreenshotError = false, title: "Close" }, h("svg", { width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round" }, h("line", { x1: "18", y1: "6", x2: "6", y2: "18" }), h("line", { x1: "6", y1: "6", x2: "18", y2: "18" })))))), this.showModal && !this.embedded && (h("div", { class: `feedback-overlay ${this.isAnimating ? 'feedback-overlay--visible' : ''}` })), this.showModal && (h("div", { class: `feedback-modal-content feedback-modal-content--${this.modalPosition} ${this.isAnimating ? 'feedback-modal-content--open' : ''} ${this.embedded ? 'feedback-modal-content--embedded' : ''}`, ref: (el) => (this.modalContent = el) }, h("div", { class: `feedback-modal-header ${(this.formSuccess && !this.successMessage) || (this.formError && !this.errorMessage) ? 'feedback-modal-header--no-content' : ''}` }, h("div", { class: "feedback-modal-header-content" }, !this.formSuccess && !this.formError ? (h("div", { class: "feedback-modal-header-text" }, h("span", { class: "feedback-modal-title" }, this.modalTitle), !this.whitelabel && (h("span", { class: "feedback-modal-powered-by" }, "Powered by", ' ', h("a", { target: "_blank", href: "https://pushfeedback.com" }, "PushFeedback"))))) : this.formSuccess ? (h("span", { class: "feedback-modal-title" }, this.modalTitleSuccess)) : (h("span", { class: "feedback-modal-title" }, this.modalTitleError))), !this.embedded && (h("button", { class: "feedback-modal-close", onClick: this.close }, h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "22", height: "22", viewBox: "0 0 24 24", fill: "none", stroke: "#191919", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", class: "feather feather-x" }, h("line", { x1: "18", y1: "6", x2: "6", y2: "18" }), h("line", { x1: "6", y1: "6", x2: "18", y2: "18" }))))), h("div", { class: "feedback-modal-body" }, !this.formSuccess && !this.formError ? (h("form", { onSubmit: this.handleSubmit }, !this.hideRating && (h("div", { class: "feedback-modal-rating" }, this.ratingMode === 'thumbs' ? (h("div", { class: "feedback-modal-rating-content" }, h("span", { class: "feedback-modal-input-heading" }, this.ratingPlaceholder), h("div", { class: "feedback-modal-rating-buttons feedback-modal-rating-buttons--thumbs" }, h("button", { title: "Yes", class: `feedback-modal-rating-button ${this.selectedRating === 1
        ? 'feedback-modal-rating-button--selected'
        : ''}`, onClick: (event) => {
        event.preventDefault();
        this.handleRatingChange(1);
      } }, h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "24", height: "24", viewBox: "0 0 24 24", fill: "none", stroke: "#5F6368", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round" }, h("path", { d: "M14 9V5a3 3 0 0 0-3-3l-4 9v11h11.28a2 2 0 0 0 2-1.7l1.38-9a2 2 0 0 0-2-2.3zM7 22H4a2 2 0 0 1-2-2v-7a2 2 0 0 1 2-2h3" }))), h("button", { title: "No", class: `feedback-modal-rating-button ${this.selectedRating === 5
        ? 'feedback-modal-rating-button--selected'
        : ''}`, onClick: (event) => {
        event.preventDefault();
        this.handleRatingChange(5);
      } }, h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "24", height: "24", viewBox: "0 0 24 24", fill: "none", stroke: "#5F6368", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round" }, h("path", { d: "M10 15v4a3 3 0 0 0 3 3l4-9V2H5.72a2 2 0 0 0-2 1.7l-1.38 9a2 2 0 0 0 2 2.3zm7-13h2.67A2.31 2.31 0 0 1 22 4v7a2.31 2.31 0 0 1-2.33 2H17" })))))) : (h("div", { class: "feedback-modal-rating-content" }, h("span", { class: "feedback-modal-input-heading" }, this.ratingStarsPlaceholder), h("div", { class: "feedback-modal-rating-buttons feedback-modal-rating-buttons--stars" }, [1, 2, 3, 4, 5].map((rating) => (h("button", { key: rating, class: `feedback-modal-rating-button ${this.selectedRating >= rating
        ? 'feedback-modal-rating-button--selected'
        : ''}`, onClick: (event) => {
        event.preventDefault();
        this.handleRatingChange(rating);
      } }, h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "28", height: "28", viewBox: "0 0 24 24", fill: "none", stroke: "#5F6368", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round" }, h("polygon", { points: "12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2" })))))))))), h("div", { class: "feedback-modal-text" }, h("textarea", { placeholder: this.messagePlaceholder, value: this.formMessage, onInput: (event) => this.handleMessageInput(event) })), !this.hideEmail && (h("div", { class: "feedback-modal-email" }, h("input", { placeholder: this.emailPlaceholder, type: "email", onInput: (event) => this.handleEmailInput(event), value: this.formEmail, required: this.isEmailRequired }))), h("div", { class: "feedback-verification" }, h("input", { type: "text", name: "verification", style: { display: 'none' }, onInput: (event) => this.handleVerification(event), value: this.formVerification })), !this.hidePrivacyPolicy && (h("div", { class: "feedback-modal-privacy" }, h("input", { type: "checkbox", id: "privacyPolicy", onChange: (ev) => this.handleCheckboxChange(ev), required: true }), h("span", { innerHTML: this.privacyPolicyText }))), h("div", { class: `feedback-modal-buttons ${this.hideScreenshotButton ? 'single' : ''}` }, !this.hideScreenshotButton && (h("button", { type: "button", class: `feedback-modal-button feedback-modal-button--screenshot ${this.encodedScreenshot ? 'feedback-modal-button--active' : ''}`, onClick: this.openScreenShot, disabled: this.sending || this.takingScreenshot }, this.encodedScreenshot && (h("div", { class: "screenshot-preview", onClick: this.openCanvasEditor }, h("img", { src: this.encodedScreenshot, alt: "Screenshot Preview" }))), !this.encodedScreenshot && !this.takingScreenshot && (h("svg", { xmlns: "http://www.w3.org/2000/svg", height: "24", viewBox: "0 -960 960 960", width: "24" }, h("path", { d: "M680-80v-120H560v-80h120v-120h80v120h120v80H760v120h-80ZM200-200v-200h80v120h120v80H200Zm0-360v-200h200v80H280v120h-80Zm480 0v-120H560v-80h200v200h-80Z" }))), this.takingScreenshot && (h("div", { class: "screenshot-loading" }, h("svg", { width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "#666", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", class: "feather-loader" }, h("line", { x1: "12", y1: "2", x2: "12", y2: "6" }), h("line", { x1: "12", y1: "18", x2: "12", y2: "22" }), h("line", { x1: "4.93", y1: "4.93", x2: "7.76", y2: "7.76" }), h("line", { x1: "16.24", y1: "16.24", x2: "19.07", y2: "19.07" }), h("line", { x1: "2", y1: "12", x2: "6", y2: "12" }), h("line", { x1: "18", y1: "12", x2: "22", y2: "12" }), h("line", { x1: "4.93", y1: "19.07", x2: "7.76", y2: "16.24" }), h("line", { x1: "16.24", y1: "7.76", x2: "19.07", y2: "4.93" })))), this.takingScreenshot ? this.screenshotTakingText :
      this.encodedScreenshot ? this.screenshotAttachedText : this.screenshotButtonText)), h("button", { class: "feedback-modal-button feedback-modal-button--submit", type: "submit", disabled: this.sending }, this.sendButtonText)))) : this.formSuccess && !this.formError ? (h("div", { class: "feedback-modal-success" }, h("p", { class: "feedback-modal-message" }, this.successMessage))) : this.formError && this.formErrorStatus == 404 ? (h("p", { class: "feedback-modal-message" }, this.errorMessage404)) : this.formError && this.formErrorStatus == 403 ? (h("p", { class: "feedback-modal-message" }, this.errorMessage403)) : this.formError ? (h("p", { class: "feedback-modal-message" }, this.errorMessage)) : (h("span", null))), !this.formSuccess && !this.formError && (this.recaptchaEnabled || this.footerText) && (h("div", { class: "feedback-modal-footer" }, h("div", { class: "feedback-footer-combined" }, this.recaptchaEnabled && (h("span", { innerHTML: this.recaptchaText })), this.recaptchaEnabled && this.footerText && ' ', this.footerText && (h("span", { innerHTML: this.footerText })))))))));
  }
  componentDidRender() {
    if (this.showModal) {
      requestAnimationFrame(() => {
        this.overlayVisible = true;
      });
    }
  }
  async openModal() {
    this.showModal = true;
    requestAnimationFrame(() => {
      requestAnimationFrame(() => {
        this.isAnimating = true;
      });
    });
  }
  static get style() { return feedbackModalCss; }
}, [1, "feedback-modal", {
    "customFont": [4, "custom-font"],
    "emailAddress": [1, "email-address"],
    "hideEmail": [4, "hide-email"],
    "isEmailRequired": [4, "is-email-required"],
    "ratingMode": [1, "rating-mode"],
    "hasSelectedElement": [1540, "has-selected-element"],
    "hidePrivacyPolicy": [4, "hide-privacy-policy"],
    "hideRating": [4, "hide-rating"],
    "hideScreenshotButton": [4, "hide-screenshot-button"],
    "project": [1],
    "showScreenshotMode": [1540, "show-screenshot-mode"],
    "showScreenshotTopBar": [1540, "show-screenshot-top-bar"],
    "showModal": [1540, "show-modal"],
    "rating": [2],
    "metadata": [1],
    "fetchData": [4, "fetch-data"],
    "embedded": [4],
    "emailPlaceholder": [1, "email-placeholder"],
    "errorMessage": [1, "error-message"],
    "errorMessage403": [1, "error-message-4-0-3"],
    "errorMessage404": [1, "error-message-4-0-4"],
    "messagePlaceholder": [1, "message-placeholder"],
    "footerText": [1, "footer-text"],
    "modalPosition": [1, "modal-position"],
    "modalTitle": [1, "modal-title"],
    "modalTitleError": [1, "modal-title-error"],
    "modalTitleSuccess": [1, "modal-title-success"],
    "privacyPolicyText": [1, "privacy-policy-text"],
    "ratingPlaceholder": [1, "rating-placeholder"],
    "ratingStarsPlaceholder": [1, "rating-stars-placeholder"],
    "sendButtonText": [1, "send-button-text"],
    "successMessage": [1, "success-message"],
    "recaptchaText": [1, "recaptcha-text"],
    "screenshotAttachedText": [1, "screenshot-attached-text"],
    "screenshotButtonText": [1, "screenshot-button-text"],
    "screenshotTakingText": [1, "screenshot-taking-text"],
    "screenshotEditTextButtonText": [1, "screenshot-edit-text-button-text"],
    "screenshotEditorTitle": [1, "screenshot-editor-title"],
    "screenshotEditorCancelText": [1, "screenshot-editor-cancel-text"],
    "screenshotEditorSaveText": [1, "screenshot-editor-save-text"],
    "screenshotSizeLabelText": [1, "screenshot-size-label-text"],
    "screenshotBorderLabelText": [1, "screenshot-border-label-text"],
    "screenshotEditTextPromptText": [1, "screenshot-edit-text-prompt-text"],
    "screenshotErrorGeneral": [1, "screenshot-error-general"],
    "screenshotErrorPermission": [1, "screenshot-error-permission"],
    "screenshotErrorNotSupported": [1, "screenshot-error-not-supported"],
    "screenshotErrorNotFound": [1, "screenshot-error-not-found"],
    "screenshotErrorCancelled": [1, "screenshot-error-cancelled"],
    "screenshotErrorBrowserNotSupported": [1, "screenshot-error-browser-not-supported"],
    "screenshotErrorUnexpected": [1, "screenshot-error-unexpected"],
    "sending": [32],
    "formMessage": [32],
    "formEmail": [32],
    "formSuccess": [32],
    "formVerification": [32],
    "formError": [32],
    "formErrorStatus": [32],
    "encodedScreenshot": [32],
    "isPrivacyChecked": [32],
    "whitelabel": [32],
    "selectedRating": [32],
    "overlayVisible": [32],
    "isAnimating": [32],
    "recaptchaEnabled": [32],
    "recaptchaSiteKey": [32],
    "takingScreenshot": [32],
    "showScreenshotError": [32],
    "screenshotError": [32],
    "showCanvasEditor": [32],
    "autoStartCapture": [32],
    "openModal": [64]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["feedback-modal", "canvas-editor"];
  components.forEach(tagName => { switch (tagName) {
    case "feedback-modal":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, FeedbackModal);
      }
      break;
    case "canvas-editor":
      if (!customElements.get(tagName)) {
        defineCustomElement$1();
      }
      break;
  } });
}

export { FeedbackModal as F, defineCustomElement as d };
