// instantsearch.js@4.87.1/es/connectors/filter-suggestions/connectFilterSuggestions.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import { checkRendering, createDocumentationMessageGenerator, getAlgoliaAgent, getAppIdAndApiKey, getRefinements, noop } from "../../lib/utils/index.js";
var withUsage = createDocumentationMessageGenerator({
  name: 'filter-suggestions',
  connector: true
});
var connectFilterSuggestions = function connectFilterSuggestions(renderFn) {
  var unmountFn = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : noop;
  checkRendering(renderFn, withUsage());
  return function (widgetParams) {
    var agentId = widgetParams.agentId,
      attributes = widgetParams.attributes,
      _widgetParams$maxSugg = widgetParams.maxSuggestions,
      maxSuggestions = _widgetParams$maxSugg === void 0 ? 3 : _widgetParams$maxSugg,
      _widgetParams$debounc = widgetParams.debounceMs,
      debounceMs = _widgetParams$debounc === void 0 ? 300 : _widgetParams$debounc,
      _widgetParams$hitsToS = widgetParams.hitsToSample,
      hitsToSample = _widgetParams$hitsToS === void 0 ? 5 : _widgetParams$hitsToS,
      _widgetParams$transfo = widgetParams.transformItems,
      transformItems = _widgetParams$transfo === void 0 ? function (items) {
        return items;
      } : _widgetParams$transfo,
      transport = widgetParams.transport;
    if (!agentId && !transport) {
      throw new Error(withUsage('The `agentId` option is required unless a custom `transport` is provided.'));
    }
    var endpoint;
    var headers;
    var suggestions = [];
    var isLoading = false;
    var debounceTimer;
    var lastStateSignature = null; // null means never fetched
    var refine;
    var searchHelper = null;
    var latestRenderOptions = null;

    // Create a signature of the current search state (query + refinements)
    var getStateSignature = function getStateSignature(results) {
      var query = results.query || '';
      var refinements = searchHelper ? JSON.stringify(searchHelper.state.facetsRefinements) + JSON.stringify(searchHelper.state.disjunctiveFacetsRefinements) + JSON.stringify(searchHelper.state.hierarchicalFacetsRefinements) : '';
      return "".concat(query, "|").concat(refinements);
    };
    var _getWidgetRenderState = function getWidgetRenderState(renderOptions) {
      var results = 'results' in renderOptions ? renderOptions.results : undefined;
      var transformedSuggestions = transformItems(suggestions, {
        results: results
      });
      return {
        suggestions: transformedSuggestions,
        isLoading: isLoading,
        refine: refine,
        widgetParams: widgetParams
      };
    };

    // Minimum duration to show skeleton to avoid flash when results are cached
    var MIN_SKELETON_DURATION_MS = 300;
    var fetchSuggestions = function fetchSuggestions(results, renderOptions) {
      var _results$hits, _rawResults$;
      if (!(results !== null && results !== void 0 && (_results$hits = results.hits) !== null && _results$hits !== void 0 && _results$hits.length)) {
        suggestions = [];
        isLoading = false;
        renderFn(_objectSpread(_objectSpread({}, _getWidgetRenderState(renderOptions)), {}, {
          instantSearchInstance: renderOptions.instantSearchInstance
        }), false);
        return;
      }
      var loadingStartTime = Date.now();
      isLoading = true;
      renderFn(_objectSpread(_objectSpread({}, _getWidgetRenderState(renderOptions)), {}, {
        instantSearchInstance: renderOptions.instantSearchInstance
      }), false);

      // Get facets from raw results (results.facets is processed differently)
      var rawResults = results._rawResults;
      var rawFacets = (rawResults === null || rawResults === void 0 ? void 0 : (_rawResults$ = rawResults[0]) === null || _rawResults$ === void 0 ? void 0 : _rawResults$.facets) || {};
      var facetsToSend = attributes ? Object.fromEntries(Object.entries(rawFacets).filter(function (_ref) {
        var _ref2 = _slicedToArray(_ref, 1),
          key = _ref2[0];
        return attributes.includes(key);
      })) : rawFacets;

      // Collect current refinements to exclude from suggestions
      var currentRefinements = searchHelper ? getRefinements(results, searchHelper.state).map(function (refinement) {
        return {
          attribute: refinement.attribute,
          value: refinement.name
        };
      }) : [];
      var messageText = JSON.stringify({
        query: results.query,
        facets: facetsToSend,
        hitsSample: results.hits.slice(0, hitsToSample),
        currentRefinements: currentRefinements,
        maxSuggestions: maxSuggestions
      });
      var payload = {
        messages: [{
          id: "sr-".concat(Date.now()),
          createdAt: new Date().toISOString(),
          role: 'user',
          parts: [{
            type: 'text',
            text: messageText
          }]
        }]
      };

      // Apply custom body transformation if provided
      var finalPayload = transport !== null && transport !== void 0 && transport.prepareSendMessagesRequest ? transport.prepareSendMessagesRequest(payload).body : payload;
      fetch(endpoint, {
        method: 'POST',
        headers: _objectSpread(_objectSpread({}, headers), {}, {
          'Content-Type': 'application/json'
        }),
        body: JSON.stringify(finalPayload)
      }).then(function (response) {
        if (!response.ok) {
          throw new Error("HTTP error ".concat(response.status));
        }
        return response.json();
      }).then(function (data) {
        var parsedSuggestions = JSON.parse(data.parts[1].text);
        var validSuggestions = (Array.isArray(parsedSuggestions) ? parsedSuggestions : []).filter(function (suggestion) {
          if (!(suggestion !== null && suggestion !== void 0 && suggestion.attribute) || !(suggestion !== null && suggestion !== void 0 && suggestion.value) || !(suggestion !== null && suggestion !== void 0 && suggestion.label)) {
            return false;
          }
          // If attributes filter is specified, only allow suggestions for those attributes
          if (attributes && !attributes.includes(suggestion.attribute)) {
            return false;
          }
          return true;
        }).slice(0, maxSuggestions);
        suggestions = validSuggestions;
      }).catch(function () {
        suggestions = [];
      }).finally(function () {
        var elapsed = Date.now() - loadingStartTime;
        var remainingDelay = Math.max(0, MIN_SKELETON_DURATION_MS - elapsed);
        var finishLoading = function finishLoading() {
          isLoading = false;
          renderFn(_objectSpread(_objectSpread({}, _getWidgetRenderState(renderOptions)), {}, {
            instantSearchInstance: renderOptions.instantSearchInstance
          }), false);
        };
        if (remainingDelay > 0) {
          setTimeout(finishLoading, remainingDelay);
        } else {
          finishLoading();
        }
      });
    };
    return {
      $$type: 'ais.filterSuggestions',
      init: function init(initOptions) {
        var instantSearchInstance = initOptions.instantSearchInstance,
          helper = initOptions.helper;
        searchHelper = helper;
        if (transport) {
          // Use custom transport configuration
          endpoint = transport.api;
          headers = transport.headers || {};
        } else {
          // Use default Algolia agent endpoint
          var _getAppIdAndApiKey = getAppIdAndApiKey(instantSearchInstance.client),
            _getAppIdAndApiKey2 = _slicedToArray(_getAppIdAndApiKey, 2),
            appId = _getAppIdAndApiKey2[0],
            apiKey = _getAppIdAndApiKey2[1];
          if (!appId || !apiKey) {
            throw new Error(withUsage('Could not extract Algolia credentials from the search client.'));
          }
          endpoint = "https://".concat(appId, ".algolia.net/agent-studio/1/agents/").concat(agentId, "/completions?compatibilityMode=ai-sdk-5&stream=false");
          headers = {
            'x-algolia-application-id': appId,
            'x-algolia-api-key': apiKey,
            'x-algolia-agent': getAlgoliaAgent(instantSearchInstance.client)
          };
        }
        refine = function refine(attribute, value) {
          var _helper$state$hierarc;
          // Check if the attribute belongs to a hierarchical facet
          // by finding a hierarchical facet that includes this attribute
          var attr = ((_helper$state$hierarc = helper.state.hierarchicalFacets.find(function (facet) {
            return facet.attributes.includes(attribute);
          })) === null || _helper$state$hierarc === void 0 ? void 0 : _helper$state$hierarc.name) || attribute;
          helper.toggleFacetRefinement(attr, value);
          helper.search();
        };
        renderFn(_objectSpread(_objectSpread({}, _getWidgetRenderState(initOptions)), {}, {
          instantSearchInstance: instantSearchInstance
        }), true);
      },
      render: function render(renderOptions) {
        var results = renderOptions.results,
          instantSearchInstance = renderOptions.instantSearchInstance;

        // Always store the latest render options
        latestRenderOptions = renderOptions;
        if (!results) {
          renderFn(_objectSpread(_objectSpread({}, _getWidgetRenderState(renderOptions)), {}, {
            instantSearchInstance: instantSearchInstance
          }), false);
          return;
        }

        // Debounce: only fetch if search state changed (query or refinements) and after delay
        var stateSignature = getStateSignature(results);
        if (stateSignature !== lastStateSignature) {
          lastStateSignature = stateSignature;
          clearTimeout(debounceTimer);
          debounceTimer = setTimeout(function () {
            var _latestRenderOptions;
            // Use the latest render options when the timeout fires
            if ((_latestRenderOptions = latestRenderOptions) !== null && _latestRenderOptions !== void 0 && _latestRenderOptions.results) {
              fetchSuggestions(latestRenderOptions.results, latestRenderOptions);
            }
          }, debounceMs);
        }
        renderFn(_objectSpread(_objectSpread({}, _getWidgetRenderState(renderOptions)), {}, {
          instantSearchInstance: instantSearchInstance
        }), false);
      },
      dispose: function dispose() {
        clearTimeout(debounceTimer);
        unmountFn();
      },
      getRenderState: function getRenderState(renderState, renderOptions) {
        return _objectSpread(_objectSpread({}, renderState), {}, {
          filterSuggestions: this.getWidgetRenderState(renderOptions)
        });
      },
      getWidgetRenderState: function getWidgetRenderState(renderOptions) {
        return _getWidgetRenderState(renderOptions);
      }
    };
  };
};
export default connectFilterSuggestions;
