// instantsearch.js@4.87.1/cjs/lib/ai-lite/transport.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HttpChatTransport = exports.DefaultChatTransport = void 0;
var _streamParser = require("./stream-parser");
var _utils = require("./utils");
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
function _createSuper(t) { var r = _isNativeReflectConstruct(); return function () { var e, o = _getPrototypeOf(t); if (r) { var s = _getPrototypeOf(this).constructor; e = Reflect.construct(o, arguments, s); } else e = o.apply(this, arguments); return _possibleConstructorReturn(this, e); }; }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /**
 * HTTP transport implementation for chat.
 */
/**
 * Abstract base class for HTTP-based chat transports.
 */
var HttpChatTransport = exports.HttpChatTransport = /*#__PURE__*/function () {
  function HttpChatTransport(_ref) {
    var _ref$api = _ref.api,
      api = _ref$api === void 0 ? '/api/chat' : _ref$api,
      credentials = _ref.credentials,
      headers = _ref.headers,
      body = _ref.body,
      customFetch = _ref.fetch,
      prepareSendMessagesRequest = _ref.prepareSendMessagesRequest,
      prepareReconnectToStreamRequest = _ref.prepareReconnectToStreamRequest;
    _classCallCheck(this, HttpChatTransport);
    _defineProperty(this, "api", void 0);
    _defineProperty(this, "credentials", void 0);
    _defineProperty(this, "headers", void 0);
    _defineProperty(this, "body", void 0);
    _defineProperty(this, "fetch", void 0);
    _defineProperty(this, "prepareSendMessagesRequest", void 0);
    _defineProperty(this, "prepareReconnectToStreamRequest", void 0);
    this.api = api;
    this.credentials = credentials;
    this.headers = headers;
    this.body = body;
    this.fetch = customFetch;
    this.prepareSendMessagesRequest = prepareSendMessagesRequest;
    this.prepareReconnectToStreamRequest = prepareReconnectToStreamRequest;
  }
  _createClass(HttpChatTransport, [{
    key: "sendMessages",
    value: function sendMessages(_ref2) {
      var _this$fetch,
        _this = this;
      var abortSignal = _ref2.abortSignal,
        chatId = _ref2.chatId,
        messages = _ref2.messages,
        requestMetadata = _ref2.requestMetadata,
        trigger = _ref2.trigger,
        messageId = _ref2.messageId,
        requestHeaders = _ref2.headers,
        requestBody = _ref2.body;
      var fetchFn = (_this$fetch = this.fetch) !== null && _this$fetch !== void 0 ? _this$fetch : fetch;

      // Resolve configurable values
      return Promise.all([(0, _utils.resolveValue)(this.credentials), (0, _utils.resolveValue)(this.headers), (0, _utils.resolveValue)(this.body)]).then(function (_ref3) {
        var _ref4 = _slicedToArray(_ref3, 3),
          resolvedCredentials = _ref4[0],
          resolvedHeaders = _ref4[1],
          resolvedBody = _ref4[2];
        // Build default request options
        var api = _this.api;
        var body = _objectSpread(_objectSpread({
          id: chatId,
          messages: messages
        }, resolvedBody), requestBody);
        var headers = _objectSpread(_objectSpread({
          'Content-Type': 'application/json'
        }, resolvedHeaders instanceof Headers ? Object.fromEntries(resolvedHeaders.entries()) : resolvedHeaders), requestHeaders instanceof Headers ? Object.fromEntries(requestHeaders.entries()) : requestHeaders);
        var credentials = resolvedCredentials;

        // Apply custom preparation if provided
        var prepareRequestBody = _objectSpread(_objectSpread({}, resolvedBody), requestBody);
        var preparePromise = _this.prepareSendMessagesRequest ? Promise.resolve(_this.prepareSendMessagesRequest({
          id: chatId,
          messages: messages,
          requestMetadata: requestMetadata,
          body: prepareRequestBody,
          credentials: resolvedCredentials,
          headers: resolvedHeaders,
          api: _this.api,
          trigger: trigger,
          messageId: messageId
        })) : Promise.resolve(null);
        return preparePromise.then(function (prepared) {
          if (prepared) {
            body = prepared.body;
            if (prepared.api) api = prepared.api;
            if (prepared.headers) {
              headers = _objectSpread({
                'Content-Type': 'application/json'
              }, prepared.headers instanceof Headers ? Object.fromEntries(prepared.headers.entries()) : prepared.headers);
            }
            if (prepared.credentials) credentials = prepared.credentials;
          }
          return fetchFn(api, {
            method: 'POST',
            headers: headers,
            body: JSON.stringify(body),
            signal: abortSignal,
            credentials: credentials
          }).then(function (response) {
            if (!response.ok) {
              throw new Error("HTTP error: ".concat(response.status, " ").concat(response.statusText));
            }
            if (!response.body) {
              throw new Error('Response body is empty');
            }
            return _this.processResponseStream(response.body);
          });
        });
      });
    }
  }, {
    key: "reconnectToStream",
    value: function reconnectToStream(_ref5) {
      var _this$fetch2,
        _this2 = this;
      var chatId = _ref5.chatId,
        requestHeaders = _ref5.headers,
        requestBody = _ref5.body;
      var fetchFn = (_this$fetch2 = this.fetch) !== null && _this$fetch2 !== void 0 ? _this$fetch2 : fetch;

      // Resolve configurable values
      return Promise.all([(0, _utils.resolveValue)(this.credentials), (0, _utils.resolveValue)(this.headers), (0, _utils.resolveValue)(this.body)]).then(function (_ref6) {
        var _ref7 = _slicedToArray(_ref6, 3),
          resolvedCredentials = _ref7[0],
          resolvedHeaders = _ref7[1],
          resolvedBody = _ref7[2];
        // Build default request options
        var api = _this2.api;
        var headers = _objectSpread(_objectSpread({}, resolvedHeaders instanceof Headers ? Object.fromEntries(resolvedHeaders.entries()) : resolvedHeaders), requestHeaders instanceof Headers ? Object.fromEntries(requestHeaders.entries()) : requestHeaders);
        var credentials = resolvedCredentials;

        // Apply custom preparation if provided
        var prepareRequestBody = _objectSpread(_objectSpread({}, resolvedBody), requestBody);
        var preparePromise = _this2.prepareReconnectToStreamRequest ? Promise.resolve(_this2.prepareReconnectToStreamRequest({
          id: chatId,
          requestMetadata: undefined,
          body: prepareRequestBody,
          credentials: resolvedCredentials,
          headers: resolvedHeaders,
          api: _this2.api
        })) : Promise.resolve(null);
        return preparePromise.then(function (prepared) {
          if (prepared) {
            if (prepared.api) api = prepared.api;
            if (prepared.headers) {
              headers = prepared.headers instanceof Headers ? Object.fromEntries(prepared.headers.entries()) : prepared.headers;
            }
            if (prepared.credentials) credentials = prepared.credentials;
          }

          // GET request for reconnection
          return fetchFn("".concat(api, "?chatId=").concat(chatId), {
            method: 'GET',
            headers: headers,
            credentials: credentials
          }).then(function (response) {
            if (!response.ok) {
              // 404 means no stream to reconnect to, which is not an error
              if (response.status === 404) {
                return null;
              }
              throw new Error("HTTP error: ".concat(response.status, " ").concat(response.statusText));
            }
            if (!response.body) {
              return null;
            }
            return _this2.processResponseStream(response.body);
          });
        });
      });
    }
  }]);
  return HttpChatTransport;
}();
/**
 * Default chat transport implementation using NDJSON streaming.
 */
var DefaultChatTransport = exports.DefaultChatTransport = /*#__PURE__*/function (_HttpChatTransport) {
  _inherits(DefaultChatTransport, _HttpChatTransport);
  var _super = _createSuper(DefaultChatTransport);
  function DefaultChatTransport() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    _classCallCheck(this, DefaultChatTransport);
    return _super.call(this, options);
  }
  _createClass(DefaultChatTransport, [{
    key: "processResponseStream",
    value: function processResponseStream(stream) {
      return (0, _streamParser.parseJsonEventStream)(stream);
    }
  }]);
  return DefaultChatTransport;
}(HttpChatTransport);
